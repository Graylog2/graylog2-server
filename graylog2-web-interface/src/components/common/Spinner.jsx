import React from 'react';
import PropTypes from 'prop-types';

/**
 * Simple spinner to use while waiting for something to load.
 */
const Spinner = ({ text }) => <span><i className="fa fa-spin fa-spinner" /> {text}</span>;

Spinner.propTypes = {
  /** Text to show while loading. */
  text: PropTypes.string,
};
Spinner.defaultProps = {
  text: 'Loading...',
};

export default Spinner;
