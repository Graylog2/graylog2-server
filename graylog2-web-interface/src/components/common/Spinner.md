Spinner with custom text:
```js
<Spinner text="Waiting for something to load..." />
```