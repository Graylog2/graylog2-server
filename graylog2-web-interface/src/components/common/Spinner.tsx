/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import * as React from 'react';
import styled, { css } from 'styled-components';

import type { IconName } from 'components/common/Icon';

import Icon from './Icon';
import Delayed from './Delayed';

const Container = styled.span`
  display: inline-flex;
  align-items: center;
  vertical-align: baseline;
`;

const StyledIcon = styled(Icon)<{ $displayMargin: boolean }>(({ $displayMargin }) => css`
  ${$displayMargin ? 'margin-right: 6px;' : ''}
`);

type Props = {
  delay?: number,
  name?: IconName,
  text?: string,
  size?: React.ComponentProps<typeof StyledIcon>['size'],
  style?: React.ComponentProps<typeof StyledIcon>['style'],
};

/**
 * Simple spinner to use while waiting for something to load.
 */
const Spinner = ({ name = 'progress_activity', text = 'Loading...', delay = 200, ...rest }: Props) => (
  <Delayed delay={delay}>
    <Container>
      <StyledIcon {...rest} name={name} $displayMargin={!!text?.trim()} spin />{text}
    </Container>
  </Delayed>
);

export default Spinner;
