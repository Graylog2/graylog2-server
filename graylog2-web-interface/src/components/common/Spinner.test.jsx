import React from 'react';
import { render, cleanup } from 'wrappedTestingLibrary';

import Spinner from 'components/common/Spinner';

describe('<Spinner />', () => {
  afterEach(() => {
    cleanup();
  });

  it('should render without props', () => {
    const { getByText } = render(<Spinner />);
    expect(getByText('Loading...')).not.toBeNull();
  });

  it('should render with a different text string', () => {
    const text = 'Hello world!';
    const { getByText } = render(<Spinner text={text} />);
    expect(getByText(text)).not.toBeNull();
  });
});
