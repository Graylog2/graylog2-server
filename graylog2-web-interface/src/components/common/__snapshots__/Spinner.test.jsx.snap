// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Spinner /> should render with a different text string 1`] = `
<Spinner
  name="spinner"
  text="Hello world!"
>
  <span>
    <ForwardRef
      fixedWidth={false}
      inverse={false}
      name="spinner"
      pulse={false}
      spin={true}
    >
      <i
        className="fa fa-spinner fa-spin"
      />
    </ForwardRef>
     
    Hello world!
  </span>
</Spinner>
`;

exports[`<Spinner /> should render without props 1`] = `
<Spinner
  name="spinner"
  text="Loading..."
>
  <span>
    <ForwardRef
      fixedWidth={false}
      inverse={false}
      name="spinner"
      pulse={false}
      spin={true}
    >
      <i
        className="fa fa-spinner fa-spin"
      />
    </ForwardRef>
     
    Loading...
  </span>
</Spinner>
`;
